compiler/rustc_query_system/src/dep_graph/graph.rs

```diff
@@ -488,15 +488,52 @@ impl<K: DepKind> DepGraph<K> {
         let prev_index = data.previous.node_to_index_opt(dep_node)?;
 
         match data.colors.get(prev_index) {
-            Some(DepNodeColor::Green(dep_node_index)) => Some((prev_index, dep_node_index)),
-            Some(DepNodeColor::Red) => None,
-            None => {
-                // This DepNode and the corresponding query invocation existed
-                // in the previous compilation session too, so we can try to
-                // mark it as green by recursively marking all of its
-                // dependencies green.
-                self.try_mark_previous_green(tcx, data, prev_index, &dep_node)
-                    .map(|dep_node_index| (prev_index, dep_node_index))
+            Some(DepNodeColor::Green(dep_node_index)) => return Some((prev_index, dep_node_index)),
+            Some(DepNodeColor::Red) => return None,
+            None => {}
+        }
+
+        let mut stack = smallvec![prev_index];
+        let _backtrace_print =
+            MarkingStack { stack: &mut stack, sess: tcx.dep_context().sess(), graph: data };
+
+        // This DepNode and the corresponding query invocation existed
+        // in the previous compilation session too, so we can try to
+        // mark it as green by recursively marking all of its
+        // dependencies green.
+        let ret = self
+            .try_mark_previous_green(tcx, data, prev_index, &dep_node, _backtrace_print.stack)
+            .map(|dep_node_index| (prev_index, dep_node_index));
+
+        // We succeeded, forego printing a backtrace.
+        std::mem::forget(_backtrace_print);
```

Comment:
Isn't this a memory leak?

Reply:
`MarkingStack` only contains three references, so forgetting it cannot create a memory leak. The backing memory of the `SmallVec` is correctly dropped since it's in a separate local. I'll add a comment.